sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,691 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/az_mqtt5.h>
+#include <azure/core/az_mqtt5_rpc.h>
+#include <azure/core/az_mqtt5_rpc_client.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/az_result.h>
+#include <azure/core/internal/az_log_internal.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mqtt_protocol.h"
+
+#include <azure/core/_az_cfg.h>
+
+static az_result root(az_event_policy* me, az_event event);
+static az_result idle(az_event_policy* me, az_event event);
+static az_result subscribing(az_event_policy* me, az_event event);
+static az_result ready(az_event_policy* me, az_event event);
+static az_result faulted(az_event_policy* me, az_event event);
+
+AZ_NODISCARD az_result _az_rpc_client_hfsm_policy_init(
+    _az_hfsm* hfsm,
+    _az_event_client* event_client,
+    az_mqtt5_connection* connection);
+
+static az_event_policy_handler _get_parent(az_event_policy_handler child_state)
+{
+  az_event_policy_handler parent_state;
+
+  if (child_state == root)
+  {
+    parent_state = NULL;
+  }
+  else if (
+      child_state == idle || child_state == subscribing || child_state == ready
+      || child_state == faulted)
+  {
+    parent_state = root;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+AZ_INLINE az_mqtt5_rpc_client_rsp_event_data _parse_response(
+    az_mqtt5_recv_data* recv_data,
+    az_mqtt5_property_binarydata* correlation_data,
+    az_mqtt5_property_stringpair* status,
+    az_mqtt5_property_stringpair* error_message,
+    az_mqtt5_property_string* content_type)
+{
+  // parse response
+  printf("Received response: %s\n", az_span_ptr(recv_data->payload));
+
+  az_mqtt5_rpc_client_rsp_event_data resp_data = { .response_payload = recv_data->payload,
```

Comment:
Stack allocation won't work: immediately after returning, this is deallocated as the stack unwinds.